mrjob/fs/gcs.py

```diff
@@ -0,0 +1,224 @@
+
+        bucket_name, base_name = parse_gcs_uri(base_uri)
+
+        # allow subdirectories of the path/glob
+        if path_glob and not path_glob.endswith('/'):
+            dir_glob = path_glob + '/*'
+        else:
+            dir_glob = path_glob + '*'
+
+
+        fields_to_return = "nextPageToken,items(id,size,bucket,name,md5Hash,etag)"
+        req = self.service.objects().list(bucket=bucket_name, fields=fields_to_return)
+        while req:
+            resp = req.execute()
+            resp_json = json.loads(json.dumps(resp))
+            for item in resp_json["items"]:
+                # We generate the item URI by adding the "gs://" prefix and snipping the "/generation"
+                # number which is at the end of the id that we retrieved.
+                uri = "%s://%s" % (scheme, item["id"].rpartition("/")[0])
+                # enforce globbing
+                if not (fnmatch.fnmatchcase(uri, path_glob) or fnmatch.fnmatchcase(uri, dir_glob)):
+                    continue
+                yield uri, long(item["size"]), item["bucket"], item["name"], item["md5Hash"], item["etag"]
+            req = self.service.objects().list_next(req, resp)
+
+    def md5sum(self, path):
+        object_list = list(self.lsWithInfo(path))
+        print object_list
+        if len(object_list) == 1:
+            return binascii.hexlify(base64.decodestring(object_list[0][4]))
+        else:
+            raise Exception("path for md5 sum doesn't resolve to single object" + path)
+
+    def _cat_file(self, filename):
+        bucket_name, object_name = parse_gcs_uri(filename)
+        req = self.service.objects().get_media(bucket=bucket_name, object=object_name)
+        fh = io.BytesIO()
+        downloader = MediaIoBaseDownload(fh, req, chunksize=1024*1024)
+        done = False
+        while not done:
+            status, done = downloader.next_chunk()
+            if status:
+                log.info("Download %d%%." % int(status.progress() * 100))
+            log.info("Download Complete for " + filename)
+        """ TODO: So, at this point we have the file content as you can see it using fh.getvalue().
+        But rather than re-implementing the decompression logic (to deal with compressed files) I'd rather
+        use the same read_file() method (in util.py) that the S3 and local versions are using.
+        But I can't figure out how to build the right file object to pass it to read_file().
+        Help appreciated.
+        """
+
+    def mkdir(self, dest):
+        """Make a directory. This does nothing on GCS because there are
+        no directories.
+        """
+        raise NotImplementedError
+
+    def exists(self, path_glob):
+        """Does the given path exist?
+
+        If dest is a directory (ends with a "/"), we check if there are
+        any files starting with that path.
+        """
+        # just fall back on ls(); it's smart
+        try:
+            paths = self.ls(path_glob)
+        except:
```

Comment:
Don't just swallow any exception here (basically, if the client raises an exception when you try to list a nonexisting path, catch that).